### ISSUE TITLE
draft doesn't allow for two code blocks next to each other

### ISSUE THREAD

**Issue by ianstormtaylor** [facebook/draft-js#452]
Right now, it seems like code blocks are handled by creating a new "block" for each line of code, instead of keeping all of the code text inside a single block. Which means that it is impossible to have two code adjacent code blocks, since they will be rendered as one.

For example, imagine wanting to have a block of `javascript` followed immediately by a block of `css`, with no extra paragraphs or lines in between:

![](http://g.recordit.co/YA8r25o9EK.gif)

The demo gets around this because it uses "unstyled" lines, so the workaround is to just chuck an extra unstyled line between them. But as soon as you try to do the same with paragraphs, you end up with an entire paragraph between them, instead of two blocks of code.

In other words, this output is impossible:

```html
<p>...</p>
<p>...</p>
<pre><code>...</code></pre>
<pre><code>...</code></pre>
<p>...</p>
```

Because you'd instead have to do this:

```html
<p>...</p>
<p>...</p>
<pre><code>...</code></pre>
<p></p>
<pre><code>...</code></pre>
<p>...</p>
```

This weirdness comes up in other places, where if you're trying to serialize code blocks, you need to make sure that you continue as long as the `i+1` block is also a code block. And it also comes up when trying to attach "syntax" data to a block, since you actually need to loop and attach it to all blocks that are next to each other.

Rendering wise, to get around it feeling like each line is a block, a separate wrapper `<pre>` tag is rendered with a `.public-DraftStyleDefault-pre` class. But this is kind of just patching over the problem.

It seems like code blocks should be handled as first-class, where lines are separated by `\n` instead of by creating separate blocks. Otherwise, the leaky abstraction ends up being a problem in lots of different places throughout the codebase.

**Comment by spicyj** [comment]
#388 could perhaps offer an avenue to improving this.

**Comment by oyeanuj** [comment]
@ianstormtaylor Does this plugin help achieve your desired behavior? https://github.com/icelab/draft-js-block-breakout-plugin

**Comment by ianstormtaylor** [comment]
@oyeanuj I don't think so, but could be missing something there. I think even with the breakout, the limitation on adjacent `code-block` blocks is hardcoded in Draft core.

Although, unrelated, that plugin is sweet, I'd implemented it internally for other use cases and will switch over to it. So, thanks!

**Comment by joshmarinacci** [comment]
Is the limitation really that you can't combine two adjacent blocks, or that when you are editing within a block pressing the 'enter' key will split the block into two, instead of just adding a newline? I wonder if that's possible to customize in the keybindings.

**Comment by so4pmaker** [comment]
On a completely side note, @ianstormtaylor, your editor looks great. I'd really like to take a look at it if it's open source.

**Comment by ianstormtaylor** [comment]
Thanks @so4pmaker! That editor itself isn't open source, nope, but I am working on [Slate](https://github.com/ianstormtaylor/slate) to make arriving at something similar to it super easy. But not that exact one 😄

**Comment by so4pmaker** [comment]
Just read through the *why?* of stale. First class plugins is very
lucrative I must say.

**Comment by ap190** [comment]
@ianstormtaylor did you ever find a work around for this?

**Comment by ianstormtaylor** [comment]
@ap190 nope, I switching to using [Slate](https://github.com/ianstormtaylor/slate).

**Comment by ahungrynoob** [comment]
I found a work around ,just use keyBindingFn;
Talk is cheap.Here is my code ,using draft-js-plugin:
```javascript
import { getDefaultKeyBinding, Modifier, EditorState } from 'draft-js';
import { getCurrentBlock } from '../utils';

export default (store) => {

    return {
        keyBindingFn: (e, { getEditorState, setEditorState }) => {
            const editorState = getEditorState();
            const { block } = getCurrentBlock(editorState);
            if (e.keyCode === 13 && block.getType() === 'code-block') {
                const newContentState = Modifier.insertText(editorState.getCurrentContent(), editorState.getSelection(), '\n');
                const newEditorState = EditorState.push(editorState, newContentState, "insert-characters");
                setEditorState(newEditorState);
                return 'add-newline';
            }
            return getDefaultKeyBinding(e);
        },
    };
};
```

**Comment by ahungrynoob** [comment]
> Is the limitation really that you can't combine two adjacent blocks, or that when you are editing within a block pressing the 'enter' key will split the block into two, instead of just adding a newline? I wonder if that's possible to customize in the keybindings.

That's a good idea.Thank you.
